rest/blip_client_test.go

```diff
@@ -750,6 +968,183 @@ func (btcRunner *BlipTestClientRunner) Collection(clientID uint32, collectionNam
+				case <-btcc.parent.rt.Context().Done():
+					return
+				default:
+				}
+				changesBatch = append(changesBatch, doc)
+				if len(changesBatch) >= changesBatchSize {
+					btcc.TB().Logf("Sending batch of %d changes", len(changesBatch))
+					proposeChangesRequest := blip.NewRequest()
+					proposeChangesRequest.SetProfile(db.MessageProposeChanges)
+
+					proposeChangesRequestBody := bytes.NewBufferString(`[`)
+					for i, change := range changesBatch {
+						if i > 0 {
+							proposeChangesRequestBody.WriteString(",")
+						}
+						activeRev := change.activeRev()
+						proposeChangesRequestBody.WriteString(fmt.Sprintf(`["%s","%s"`, change.id, activeRev.version.RevID))
+						// write last known server version to support no-conflict mode
+						if serverVersion, ok := btcc.getLastReplicatedRev(change.id); ok {
+							btcc.TB().Logf("specifying last known server version for doc %s = %v", change.id, serverVersion)
+							proposeChangesRequestBody.WriteString(fmt.Sprintf(`,"%s"`, serverVersion.RevID))
+						}
+						proposeChangesRequestBody.WriteString(`]`)
+					}
+					proposeChangesRequestBody.WriteString(`]`)
+					proposeChangesRequestBodyBytes := proposeChangesRequestBody.Bytes()
+					proposeChangesRequest.SetBody(proposeChangesRequestBodyBytes)
+
+					btcc.TB().Logf("proposeChanges request: %s", string(proposeChangesRequestBodyBytes))
+
+					btcc.addCollectionProperty(proposeChangesRequest)
+
+					if err := btcc.sendPushMsg(proposeChangesRequest); err != nil {
+						btcc.TB().Errorf("Error sending proposeChanges: %v", err)
+						return
+					}
+
+					proposeChangesResponse := proposeChangesRequest.Response()
+					rspBody, err := proposeChangesResponse.Body()
+					if err != nil {
+						btcc.TB().Errorf("Error reading proposeChanges response body: %v", err)
+						return
+					}
+					errorDomain := proposeChangesResponse.Properties["Error-Domain"]
+					errorCode := proposeChangesResponse.Properties["Error-Code"]
+					if errorDomain != "" && errorCode != "" {
+						btcc.TB().Errorf("error %s %s from proposeChanges with body: %s", errorDomain, errorCode, string(rspBody))
+						return
+					}
+
+					btcc.TB().Logf("proposeChanges response: %s", string(rspBody))
+
+					var serverDeltas bool
+					if proposeChangesResponse.Properties[db.ChangesResponseDeltas] == "true" {
+						btcc.TB().Logf("server supports deltas")
+						serverDeltas = true
+					}
+
+					var response []int
+					err = base.JSONUnmarshal(rspBody, &response)
+					require.NoError(btcc.TB(), err)
+					for i, change := range changesBatch {
+						var status int
+						if i >= len(response) {
+							// trailing zeros are removed - treat as 0 from now on
+							status = 0
+						} else {
+							status = response[i]
+						}
+						switch status {
+						case 0:
+							// FIXME: `change.activeRev()` could change if the test is writing concurrently. We'll have to store the specific version we sent in proposeChanges and iterate on that.
```

Comment:
this may be a source of flakiness but not sure it will actually surface itself unless there's a test that is writing a lot of revs on the same doc whilst pushing

Reply:
in the middle of fixing this race, since it's loosely related to other race detector failures